README.md
# Juno Reverse Geocoder

This is a set of SQL scripts working on a PostGIS database that for a given
location return a structured address: from a house number up to a country.
Together this makes for a reverse geocoder, which Juno Inc. uses internally.

## Installation

Starting it should be simple:

* Install PostgreSQL, PostGIS, and osm2pgsql.
* Create a `geocoder` database in PostgreSQL.
* Do `create extension postgis` on the database.
* Download an OpenStreetMap extract for your area.
* Run `run.sh <file.osm.pbf>` to upload the area into the database and start the geocoder.

### Docker container

Check docker container [section](./docker/README.md)

## Usage

The REST API is generally compatible with
[Nominatim's](https://nominatim.org/release-docs/develop/api/Reverse/).
No details or language modifiers work. Only `jsonv2` format
is supported (no need to specify it). Coordinates and identifiers
are output as strings.

Sample call (without language):

    http://localhost:5000/reverse?lat=50.2537&lon=7.0099

Sample multilanguage response:

```json
{
    "type": "building",
    "osm_type": "way",
    "osm_id": 499933205,
    "address": {
        "road": {
            "def": "Kirchstra\u00dfe"
        },
        "house_number": "2a",
        "country": {
            "de": "Deutschland",
            "en": "Germany",
            "es": "Alemania",
            "fr": "Allemagne",
            "pl": "Niemcy",
        },
        "state": {
            "de": "Rheinland-Pfalz",
            "en": "Rhineland-Palatinate",
            "es": "Renania-Palatinado",
            "fr": "Rh\u00e9nanie-Palatinat",
            "prefix": "Bundesland"
        },
        "county": {
            "def": "Landkreis Vulkaneifel"
        },
        "village": {
            "def": "Uersfeld"
        },
        "township": {
            "def": "Kelberg",
            "prefix": "Verbandsgemeinde"
        },
        "municipality": {
            "def": "Uersfeld"
        }
    },
    "lon": "7.0097705",
    "lat": "50.25373089939598",
    "name": {},
    "display_name": "2a Kirchstra\u00dfe"
}
```

Call with language:

    http://localhost:5000/reverse?lat=50.2537&lon=7.00996&lang=en"
    
```json
{
    "type": "building",
    "osm_type": "way",
    "osm_id": 499933205,
    "address": {
        "road": "Kirchstra\u00dfe",
        "house_number": "2a",
        "country": "Germany",
        "state": "Rhineland-Palatinate",
        "county": "Landkreis Vulkaneifel",
        "village": "Uersfeld",
        "township": "Kelberg",
        "municipality": "Uersfeld"
    },
    "lon": "7.0097705",
    "lat": "50.25373089939598",
    "display_name": "2a Kirchstra\u00dfe"
}
```

Response will be translated to the desired language (if the translation exists in the source OSM files).

WARNING: for use this feature you need to re-load the OSM into the database. This version is not compatible with the old DB.


Note that non-address tags are dropped, and the only information that remains
is whether a point is a POI or not. The `type` can be one of `admin`, `road`,
`building`, `address`, or `poi`.

### Error Messages

When there is an error in URL arguments (e.g. non-numeric coordinate or `osm_id`),
you'll get HTTP 400 with a json containing a single `"message"` key.

When an object has not been found, you'll get HTTP 404 with a json containing a single
`"error"` key.

Server errors obviously lead to HTTP 500 errors, although these should not happen.

## Contributing

The reverse geocoder is basically two sets of SQL scripts inside the `sql` directory:
one for preprocessing an osm2pgsql database (`sql/prepare`) and another for
doing queries (`sql/query`). If you aim to improve the quality, go there.

Other scripts and files are merely interfaces to use and access a database with
these scripts. They should just work.

Tests reside in the `tests` directory and are written with
[behave](https://behave.readthedocs.io/en/latest/). Look into `tests/features` to get
human-readable test cases. Feel free to add tests if you think they should cover
more cases. Have a look at tests marked `@wip`: maybe you'll have an idea how to
update SQL scripts to accomodate these.

At the moment tests rely on actual OpenStreetMap data
which might be not great, since OSM is being constantly improved, so some of these
cases would get simpler with time. A good task would be to extract data for each of
these tests into a separate osm.pbf file and make the testing framework install
a separate instance of a database.

## Author and License

The geocoder was written by Ilya Zverev for Juno, published under the Apache 2.0 License.


CHANGELOG.md
# Juno Reverse Geocoder Changelog

## 1.1.0 (2023-12-15)

* Added FastAPI server, should be faster. Not in Docker yet.
* Fixed a thing in the old geocoder for recent flask-restful library.

## 1.0.1 (2020-03-13)

* Updated documentation to reflect the
  [password policy change](https://github.com/docker-library/postgres/issues/681)
  in the upstream postgres image.
* Fixed possible issue with PostgreSQL not started before using it.

## 1.0.0 (2019-11-20)

The first public release. Differences from the initially published version:

* Docker image.
* Postal codes in response.
* Integration tests were converted to stand-alone unit tests.
* 88 tests for counties in and around New York state.
* Testing via GitHub actions.


web/geocoder_fast.py
import asyncio
from . import config
from contextlib import asynccontextmanager
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from psycopg_pool import AsyncConnectionPool


app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=['*'],
)
pool = AsyncConnectionPool(
    kwargs={
        'host': config.PG_HOST,
        'port': config.PG_PORT,
        'dbname': config.PG_DATABASE,
        'user': config.PG_USER,
    },
    open=False,
)


async def check_connections():
    while True:
        await asyncio.sleep(600)
        await pool.check()


@asynccontextmanager
async def get_cursor():
    async with pool.connection() as conn:
        cursor = conn.cursor()
        try:
            yield cursor
        finally:
            await cursor.close()


@app.on_event('startup')
async def startup():
    await pool.open()
    asyncio.create_task(check_connections())


@app.get('/')
async def root():
    return {'name': 'JRG', 'version': '1.1.0'}


def pack_response(result):
    if not result:
        return None
    data = {
        'type': result[0],
        'osm_type': result[1],
        'osm_id': result[2],
        'address': {
            'road': result[3],
            'house_number': result[4],
            'postcode': result[5],
        },
        'lon': None if result[6] is None else str(result[6]),
        'lat': None if result[7] is None else str(result[7]),
        'name': result[8],
    }
    for k, v in data['address'].items():
        if v and ';' in v:
            data['address'][k] = v.split(';', 1)[0]
    return data


async def closest_object(lon, lat):
    async with get_cursor() as cur:
        await cur.execute("select * from geocode_poi(%s::numeric, %s::numeric)", (lon, lat))
        return pack_response(await cur.fetchone())


async def object_info(osm_type, osm_id):
    async with get_cursor() as cur:
        await cur.execute(
            "select * from osm_lookup(%s, %s)", (osm_type, osm_id))
        return pack_response(await cur.fetchone())


async def q_address(lon, lat):
    result = {}
    obj = None
    async with get_cursor() as cur:
        await cur.execute("select * from geocode_admin(%s::numeric, %s::numeric)", (lon, lat))
        async for row in cur:
            result[row[0]] = row[1]
            obj = {
                'osm_type': row[2],
                'osm_id': row[3],
                'lon': str(row[4]),
                'lat': str(row[5]),
            }
    if obj and obj['osm_id'] == 0:
        obj = None
    return result, obj


def make_display_name(obj):
    if obj['address'].get('road') is not None:
        if obj['address'].get('house_number') is not None:
            return '{} {}'.format(
                obj['address']['house_number'],
                obj['address']['road'])
        return obj['address']['road']
    return obj.get('name')


def prune_dict(obj):
    for k in list(obj.keys()):
        if obj[k] is None:
            del obj[k]
        elif isinstance(obj[k], dict):
            prune_dict(obj[k])


@app.get('/reverse')
async def get(lon: float | None = None, lat: float | None = None,
              osm_id: int | None = None, osm_type: str | None = None,
              admin: bool = True):
    if (lon is not None or lat is not None) and (osm_id or osm_type):
        return {'error': 'Please use either coordinates or osm id'}, 400
    if lon is not None and lat is not None:
        obj = await closest_object(lon, lat) or {}
    elif osm_id and osm_type:
        obj = await object_info(osm_type, osm_id) or {}
        if not obj or not obj.get('osm_id'):
            return {'error': 'Unable to geocode'}, 404
    else:
        return {'error': 'Missing request arguments'}, 400

    if 'address' not in obj:
        obj['address'] = {}
    if admin != 0:
        if lon is not None:
            address, backup_osm = await q_address(lon, lat)
            if backup_osm and obj.get('osm_type') is None:
                obj.update(backup_osm)
        elif obj.get('lon') is not None:
            address, _ = await q_address(float(obj['lon']), float(obj['lat']))
        else:
            address = None
        obj['address'].update(address or {})
        if 'type' not in obj:
            obj['type'] = 'admin'

    if obj.get('osm_type') is None:
        return {'error': 'Unable to geocode'}, 404

    obj['display_name'] = make_display_name(obj)
    prune_dict(obj)
    return obj
